dll/win32/kernel32/client/file/rw.c

```diff
@@ -95,16 +95,14 @@ WriteFile(IN HANDLE hFile,
             if (NT_SUCCESS(Status)) Status = Iosb.Status;
         }
 
-        if (NT_SUCCESS(Status))
-        {
-            /*
-             * lpNumberOfBytesWritten must not be NULL here, in fact Win doesn't
-             * check that case either and crashes (only after the operation
-             * completed).
-             */
-            *lpNumberOfBytesWritten = Iosb.Information;
-        }
-        else
+        /*
+         * lpNumberOfBytesWritten must not be NULL here, in fact Win doesn't
+         * check that case either and crashes (only after the operation
+         * completed).
+         */
+        *lpNumberOfBytesWritten = Iosb.Information;
```

Comment:
Should this always happen or only in the success case and when more data is available?

I would not expect another failure than (more data available) to result in a positive `lpNumberOfBytesWritten`

Reply:
I am not sure how to answer that question. 

As Simone said in the [ticket](https://jira.reactos.org/browse/CORE-17376?focusedId=141760&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-141760), the reference code into wine (https://gitlab.winehq.org/wine/wine/-/blob/master/dlls/kernelbase/file.c#L3574) is doing a simple task, that might be the same that we could apply to WriteFile (not necessary for the Calibre ticket).

He said: _"In that branch, since overlapped is evaluated before the status in ReactOS code, theorically the result / lpNumberOfBytesWritten should be always set due to the overlapped = 0 if it has to mimic wine."_